tests/php/src/DevTools/LikelyCulpritDetectorTest.php

```diff
@@ -0,0 +1,281 @@
+					],
+					'mu-plugin',
+					'bad-mu-plugin.php',
+				],
+
+				'parent theme' => [
+					[
+						ABSPATH . '/wp-includes/some-file.php', // Core is skipped.
+						__FILE__, // AMP plugin is skipped.
+						get_template_directory() . '/functions.php', // <== Likely culprit.
+						WP_PLUGIN_DIR . '/bad-plugin/bad-plugin.php',
+						WP_CONTENT_DIR . '/mu-plugins/bad-mu-plugin.php',
+						get_stylesheet_directory() . '/functions.php',
+					],
+					'theme',
+					'default',
+				],
+
+				'child theme'  => [
+					[
+						ABSPATH . '/wp-includes/some-file.php', // Core is skipped.
+						__FILE__, // AMP plugin is skipped.
+						get_stylesheet_directory() . '/functions.php', // <== Likely culprit.
+						WP_PLUGIN_DIR . '/bad-plugin/bad-plugin.php',
+						WP_CONTENT_DIR . '/mu-plugins/bad-mu-plugin.php',
+						get_template_directory() . '/functions.php',
+					],
+					'theme',
+					'default',
+				],
+			]
+		);
+	}
+
+	/**
+	 * Tests LikelyCulpritDetector::analyze_exception
+	 *
+	 * @dataProvider single_step_trace_data()
+	 *
+	 * @param string[] $file_stack    Stack of file paths.
+	 * @param string   $expected_type Expected source type.
+	 * @param string   $expected_name Expected source name.
+	 *
+	 * @covers ::analyze_exception
+	 */
+	public function test_analyze_exception( $file_stack, $expected_type, $expected_name ) {
+		$trace     = $this->get_trace_from_file_stack( $file_stack );
+		$exception = new RuntimeException();
+		$this->set_private_property(
+			$exception,
+			'file',
+			array_shift( $trace )['file']
+		);
+
+		// The trace of the exception cannot be set as it is defined by a final
+		// method and not stored as a property. Therefore, we can only test
+		// the first "file" that is encountered.
+
+		$source = $this->likely_culprit_detector->analyze_exception( $exception );
+
+		$this->assertEquals( $expected_type, $source['type'] );
+		$this->assertEquals( $expected_name, $source['name'] );
+	}
+
+	/**
+	 * Tests LikelyCulpritDetector::analyze_trace
+	 *
+	 * @dataProvider multi_step_trace_data()
+	 *
+	 * @param string[] $file_stack    Stack of file paths.
+	 * @param string   $expected_type Expected source type.
+	 * @param string   $expected_name Expected source name.
+	 *
+	 * @covers ::analyze_trace
+	 */
+	public function test_analyze_trace( $file_stack, $expected_type, $expected_name ) {
+		$source = $this->likely_culprit_detector->analyze_trace(
+			$this->get_trace_from_file_stack( $file_stack )
+		);
+
+		$this->assertEquals( $expected_type, $source['type'] );
+		$this->assertEquals( $expected_name, $source['name'] );
+	}
+
+	/**
+	 * Convert a file stack into a (simplified) trace.
+	 *
+	 * @param string[] $file_stack File stack to convert.
+	 * @return array[] Associative of (simplified) trace data.
+	 */
+	private function get_trace_from_file_stack( $file_stack ) {
+		return array_map(
+			static function ( $file ) {
+				return [ 'file' => $file ];
```

Comment:
This could also be:

```php
return compact(  'file );
```

I know `extract()` is bad. Any reason to not use `compact()`?

Reply:
I opted to go with a simple array for performance reasons. The gain in simplicity is not worth the performance impact for a single variable. `compact()` does a lot of processing, first of all because you pass in a string name instead of using the variable directly. This needs to be parsed and mapped first to be able to assign.

If you have multiple variables (or a dynamic set) though, `compact()` is certainly fine and doesn't have the security implications of `extract()`.